packages/cypress-commands/src/setups/enableNetworkShim/captureRequests.js

```diff
@@ -0,0 +1,135 @@
+import {
+    getApiBaseUrl,
+    getFullTestName,
+    splitHostAndPath,
+    toJsonBlob,
+    isPathStaticResource,
+    findMatchingRequestStub,
+} from './utils.js'
+
+export default function captureRequests(state) {
+    state.config.hosts.forEach(host => {
+        cy.intercept(host, request => {
+            request.reply(response => {
+                try {
+                    captureRequest(state, request, response)
+                } catch (error) {
+                    console.error('NetworkShim capture error', error)
+                }
+            })
+        })
+    })
+}
+
+async function captureRequest(state, request, response) {
+    const { host, path } = splitHostAndPath(request.url, state.config.hosts)
+    if (!host) {
+        // pass through
+        return response
+    }
+
+    state.count++
+
+    const testName = getFullTestName()
+    const isStaticResource = isPathStaticResource(path, state.config)
+    const requestStub = findMatchingRequestStub(
+        {
+            path,
+            method: request.method,
+            testName,
+            requestBody: request.body,
+            isStaticResource,
+        },
+        state.requests
+    )
+    const { size, text } = await toJsonBlob(response.body)
+    const scrubbedText = removeApiEndpointFromResponseBodyBlob(text)
+
+    if (requestStub) {
+        // Repeated request
+        processDuplicatedRequestStub({
+            state,
+            requestStub,
+            newResponseBody: scrubbedText,
+            responseStatus: response.statusCode,
+        })
+    } else {
+        // New request
+        state.requests.push({
+            path,
+            testName: isStaticResource ? null : testName,
+            static: isStaticResource,
+            count: 1,
+            nonDeterministic: false,
+            method: request.method,
+            requestBody: request.body,
+            requestHeaders: request.headers,
+            statusCode: response.statusCode,
+            responseBody: scrubbedText,
+            responseSize: size,
+            responseHeaders: clearDateFromResponseHeaders(response.headers),
+        })
+        state.totalResponseSize += size
+    }
+    return response
+}
+
+function processDuplicatedRequestStub({
+    state,
+    requestStub,
+    newResponseBody,
+    responseStatus,
+}) {
+    requestStub.count += 1
+    state.duplicates += 1
+    const isNotModified = responseStatus === 304
+
+    if (!requestStub.nonDeterministic) {
+        // if requestStub.responseBody equals newResponseBody this is a simple
+        // duplicated request and we don't take any action
+        // And we deal with 304s the same way
+
+        if (requestStub.responseBody !== newResponseBody && !isNotModified) {
```

Comment:
I am confused about this condition.
In order to get to this line, a matching/existing response has to be found (in `findMatchingRequestStub.js`). Inside that function, there's a condition that compares the response bodies and only returns true if they're equal.
How can this case be happening then?

Reply:
A valid question, as this is one of the more complex parts of the beast. I'll try to explain:
- We start by looking if a matching request stub is already present. That means we find a match on `path`, `method` and `requestBody`
- But once we have found this we need to look further because actually three things could be happening here:
    - The response body matches the response body (or possibly one of the response bodies) on the stub.: this simply is a repeating request.
    - The response body does not match the stub: this means the request is _nonDeterministic_, an repeating request does not yield the same response body (I will add docs with definitions).
    - The response body is empty and the response status is 304: this is the server telling the browser that it should return its cached response. So this is a repeating request, but we threat this as an exceptional case: if we were to allow the normal logic to process the response body it would conclude that the response has changed (i.e. from some sort of array/object/string into an empty string). So it would essentially be treated as a nonDeterministic response, while it is the exact opposite. So for 304 responses we basically need to just look at last know response body and use that.

I hope this helps. It's not just a difficult part to understand but also to explain 😊 .

Here's some info about 304s:
> A 304 response cannot contain a message-body; it is always terminated by the first empty line after the header fields.

Reply:
> Inside that function, there's a condition that compares the *response bodies* and only returns true if they're equal.

BTW: the highlighted part is incorrect: `findMatchingRequestStub` does not look at matching _response_ bodies, it looks at matching _request_ bodies.

Reply:
Ok, got it! thanks, now I also know, what nonDeterministic means